Feat improve local devex

osintalex opened the pull request:
Addresses a few issues I encountered while getting everything spun up for local dev before tackling first issues:

- A few things I couldn't find in the docs when getting the UI set up to run locally with yarn
- How to get hot reloading on the backend API while keeping it in kubernetes.

I don't think it would be too much work from here to get remote debugging on the k8s pod running the API set up too. Skaffold has pretty good integrations on [that](https://skaffold.dev/docs/workflows/debug/).

Happy to also take a look at that if there is interest.

osintalex committed:
feat: skaffold yaml working

osintalex committed:
feat: backend hot reload with skaffold

osintalex committed:
chore: add trailing newlines

tw4l commented on skaffold.yaml:
```diff
@@ -0,0 +1,46 @@
+      - name: btrix
+        chartPath: chart
+        valuesFiles:
+          - chart/values.yaml
```
Since this is for local development, we may also want to add `chart/local.yaml` here, since our local development documentation instructs user to add development overrides in that file rather than in `chart/values.yaml` specifically.

`./btrix setup` from the repo root directory will copy the example local config file to the expected `chart/local.yaml`, perhaps there's a way to automate creating it if it doesn't yet exist so that Skaffold doesn't complain. At the least, we could point out that option in the documentation as a step in the new (excellent!) hot reloading docs.

osintalex replied:
Makes sense! Just added the `local.yaml` file to skaffold config. 

I think that removes the need for changes to `./btrix setup` but lmk if I misunderstood you.

osintalex committed:
feat: add debugger and prioritise local values yaml

osintalex commented:
Added in interactive debugging with VSCode in last commit too. Didn't really work out of the box with Skaffold. It is more aimed than I realised at doing everything in GCP (requires you to sign into a GCP account even to debug on a local cluster :cry:). So I dropped that in favour of `debugpy`, which imo is preferable anyway since it's open source.

tw4l commented:
> Added in interactive debugging with VSCode in last commit too. Didn't really work out of the box with Skaffold. It is more aimed than I realised at doing everything in GCP (requires you to sign into a GCP account even to debug on a local cluster 😢). So I dropped that in favour of `debugpy`, which imo is preferable anyway since it's open source.

Sounds great, thank you! Will do some local testing tomorrow :) Appreciate this contribution!

vnznznz commented:
Tried this for a bit, being able to debug the backend code is great! Unfortunately I was unable to get the`op` container in the `backend` pod to auto-update too. I also encountered some weird internal k8s 404 errors when updating files in `chart/app-templates` requiring me to reset the cluster. Also tried building `frontend` with skaffold, had to enable [useDockerCLI](https://skaffold.dev/docs/references/yaml/#build-local-useDockerCLI) to get it to work.

osintalex commented:
Sorry for slow reply. What's the expected behaviour here, do you want file sync and debuggability on the op container too? I didn't add any config for that.

I also didn't add any config for the frontend since it looked much easier to me to just run it locally with yarn and then hit the backend after portforwarding to localhost. I haven't tried this, but my get is that it will be hard to get hot reload with the frontend unless you add a different dev Dockerfile that doesn't use a multistage build.

I think expected behaviour of file sync + debugger on the op container while running frontend locally and then hitting a port-forwarded backend would probably be the best expected behaviour. Lmk if I'm missing anything though.
On Tuesday, June 4th, 2024 at 7:59 AM, Vinzenz Sinapius ***@***.***> wrote:

> Tried this for a bit, being able to debug the backend code is great! Unfortunately I was unable to get theop container in the backend pod to auto-update too. I also encountered some weird internal k8s 404 errors when updating files in chart/app-templates requiring me to reset the cluster. Also tried building frontend with skaffold, had to enable [useDockerCLI](https://skaffold.dev/docs/references/yaml/#build-local-useDockerCLI) to get it to work.
>
> —
> Reply to this email directly, [view it on GitHub](https://github.com/webrecorder/browsertrix/pull/1827#issuecomment-2146754412), or [unsubscribe](https://github.com/notifications/unsubscribe-auth/APHRW4T7I7MEOIVFXRO2UODZFVQVFAVCNFSM6AAAAABILQ4O6CVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDCNBWG42TINBRGI).
> You are receiving this because you authored the thread.Message ID: ***@***.***>

vnznznz commented:
> What's the expected behaviour here, do you want file sync and debuggability on the op container too? I didn't add any config for that.

Since they share a codebase I think they should update in sync. Several times while testing skaffold I got in this situation where I tried to troubleshoot an error and didn't realize my changes weren't applied :sweat_smile: 

Having frontend in there is not necessary imo.

ikreymer commented:
@osintalex Thank you so much for your work on this! Agree with @vnznznz, let's add the op container reloading as well, otherwise it'll be a bit confusing to work on the backend if they get out of sync. No need for frontend, as we already have local loading with yarn for that.

osintalex committed:
feat: debug and hot reload for op container

osintalex committed:
chore: add trailing newline

osintalex commented:
Makes sense! I've added hot reload and debugger support for the op container now ☺️

osintalex commented:
Bump :-)

osintalex committed:
refactor: remove probes for easier local dev
